codex/sales.nim

```diff
@@ -120,27 +132,267 @@ proc load*(sales: Sales) {.async.} =
+proc onReservationAdded(sales: Sales, availability: Availability) {.async.} =
+  ## Query last 256 blocks for new requests, adding them to the queue. `push`
+  ## checks for availability before adding to the queue. If processed, the
+  ## sales agent will check if the slot is free.
+  let context = sales.context
+  let market = context.market
+  let queue = context.slotQueue
+
+  logScope:
+    topics = "sales onReservationAdded callback"
 
-  proc onRequest(requestId: RequestId, ask: StorageAsk) {.gcsafe, upraises:[].} =
-    sales.handleRequest(requestId, ask)
+  trace "reservation added, querying past storage requests to add to queue"
 
   try:
-    sales.subscription = some await sales.context.market.subscribeRequests(onRequest)
+    let events = await market.queryPastStorageRequests(256)
+    let requests = events.map(event =>
+      SlotQueueItem.init(event.requestId, event.ask, event.expiry)
+    )
+
+    trace "found past storage requested events to add to queue",
+      events = events.len
+
+    for slots in requests:
+      if err =? (await queue.push(slots)).errorOption:
+        raise err
+  except QueueNotRunningError:
+    warn "cannot push items to queue, queue is not running"
+  except NoMatchingAvailabilityError:
+    info "slot in queue had no matching availabilities, ignoring"
+  except SlotsOutOfRangeError:
+    warn "Too many slots, cannot add to queue"
+  except SlotQueueItemExistsError:
+    trace "item already exists, ignoring"
+    discard
   except CatchableError as e:
-    error "Unable to start sales", msg = e.msg
+    warn "Error adding request to SlotQueue", error = e.msg
+    discard
 
-proc stop*(sales: Sales) {.async.} =
-  if subscription =? sales.subscription:
-    sales.subscription = market.Subscription.none
+proc onStorageRequested(sales: Sales,
+                        requestId: RequestId,
+                        ask: StorageAsk,
+                        expiry: UInt256) =
+
+  logScope:
+    topics = "sales onStorageRequested"
+    requestId
+    slots = ask.slots
+    expiry
+
+  let slotQueue = sales.context.slotQueue
+
+  trace "storage requested, adding slots to queue"
+
+  without items =? SlotQueueItem.init(requestId, ask, expiry).catch, err:
+    if err of SlotsOutOfRangeError:
+      warn "Too many slots, cannot add to queue"
+    else:
+      warn "Failed to create slot queue items from request", error = err.msg
+
+  # TODO: return Future and cancel on `.stop`?
+  for item in items:
+    slotQueue.push(item).catch(
+      proc(err: ref CatchableError) =
+        if err of NoMatchingAvailabilityError:
+          info "slot in queue had no matching availabilities, ignoring"
+        elif err of SlotQueueItemExistsError:
+          error "Failed to push item to queue becaue it already exists"
+        elif err of QueueNotRunningError:
+          warn "Failed to push item to queue becaue queue is not running"
+        else:
+          warn "Error adding request to SlotQueue", error = err.msg
+    )
```

Comment:
I'm probably missing something, but can't you achieve the same behaviour with `asyncSpawn` and regular exception handling?

```suggestion
proc onStorageRequestedImpl(sales: Sales,
                        requestId: RequestId,
                        ask: StorageAsk,
                        expiry: UInt256) {.async.} =

  logScope:
    topics = "sales onStorageRequested"
    requestId
    slots = ask.slots
    expiry

  let slotQueue = sales.context.slotQueue

  trace "storage requested, adding slots to queue"

  without items =? SlotQueueItem.init(requestId, ask, expiry).catch, err:
    if err of SlotsOutOfRangeError:
      warn "Too many slots, cannot add to queue"
    else:
      warn "Failed to create slot queue items from request", error = err.msg

  # TODO: return Future and cancel on `.stop`?
  for item in items:
    try:
      await slotQueue.push(item)
    except NoMatchingAvailabilityError as err:
       info "slot in queue had no matching availabilities, ignoring"
    except SlotQueueItemExistsError as err:
       error "Failed to push item to queue becaue it already exists"
    except QueueNotRunningError as err:
       warn "Failed to push item to queue becaue queue is not running"
    except CatchableError:
       warn "Error adding request to SlotQueue", error = err.msg

proc onStorageRequested(sales: Sales,
                        requestId: RequestId,
                        ask: StorageAsk,
                        expiry: UInt256) =
  asyncSpawn onStorageRequestedImpl(salse, requestId, ask, expiry)
```

Reply:
Sure it could be done that way. Most of the benefit is putting async code in a synchronous code flow context. This is most obvious when we need to get a result from the async procedure (or at least wait until completion), in which case we'd use `.then`.

Reply:
The Promise API helps because it allows async code to be executed in a synchronous context, whereas `asyncSpawn` is a "fire and forget" in the context of the closure. For example

```nim
proc asyncProc() {.async.} =
  await sleepAsync(1.seconds)
  echo "async proc done"

proc syncProc() =
  asyncSpawn asyncProc()
  echo "sync proc done"
```

Reply:
> Using the promise api, we don't need the extra proc, nor do we need to pass closure callbacks or variables. We can also handle uncaught errors inline. We can handle everything within the context of the synchronous proc, and also have access to its closure

You can do that with asyncSpawn as well :smile: 

```nim
proc fetchFromNetwork(): Future[int] {.async.} =
  # simulate some network value fetch
  await sleepAsync(1.seconds)
  return 1

proc sync(onComplete: proc(val: int): void) =
  let someClosureVar = 2

  proc fetcher {.async.} =
    try:
      onComplete(someClosureVar + await fetchFromNetwork())
    except CatchableError as e:
      error "error doing sync stuff", error = e.msg

  asyncSpawn fetcher()

let printNetworkResult = proc(val: int): string = $val

echo sync(printNetworkResult) # prints "3"
```

Reply:
There's still an extra proc there `fetcher` inside of the `sync` proc so you can encapsulate the closure, which becomes messy when it's a big proc (as in the sales module).

Reply:
At the end of the of the day, we don't have to use `.then().catch()` (obviously), but i thought it was a nice addition that helped make the code readable without jumping through too many procs, and I started reaching for it the more times I encountered a situation in a sync proc that needed to call async code.